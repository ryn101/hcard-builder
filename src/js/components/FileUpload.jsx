import React from 'react';
import PropTypes from 'prop-types';

function FileUpload({ children, focusClass, onFileSelection }) {
    return (
        <span className="file-upload-wrapper">
            <input id="avatarFile" onChange={onFileSelection} type="file" accept="image/*" />
            <label htmlFor="avatarFile" className={focusClass}>{ children }</label>
        </span>
    );
}

FileUpload.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]).isRequired,
    focusClass: PropTypes.string,
    onFileSelection: PropTypes.func.isRequired,
};

FileUpload.defaultProps = {
    focusClass: 'focusable',
};

export default FileUpload;
