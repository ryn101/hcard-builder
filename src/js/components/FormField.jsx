import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

/**
 * Component that acts as a wrapper for the creation of a label and it's associated input field.
 * This component has two forms:
 *
 * - Shorthand - Allows the creation of the component without the need to specify a
 *               input as a child as this is auto generated. When using this form,
 *               the two supported properties of `onFieldChange` and `type` will be applied
 *               directly on the input.
 * - Nested - Used for finer attribute control of the input. This requires an explicit
 *            declaration of an input as a child of this component.
 *
 * In both forms, the fieldId will be applied to the input as it's id.
 */
function FormField({
    fieldId, as, label, children, type, onFieldChange, className,
}) {
    const FormFieldTag = as;

    return (
        <FormFieldTag className={classNames('form__field', className)}>
            {
                // Optionally render the label if the property is provided
                label ? <label className="form__label" htmlFor={fieldId}>{label}</label> : ''
            }
            {
                children
                    // If a child input element is provided, then set the field id
                    ? React.cloneElement(children, { id: fieldId })
                    : <input id={fieldId} type={type} onChange={onFieldChange} />
            }
        </FormFieldTag>
    );
}

FormField.propTypes = {
    fieldId: PropTypes.string.isRequired,
    as: PropTypes.oneOf(['div', 'li', 'p']),
    label: PropTypes.string,
    children: PropTypes.element,
    type: PropTypes.string,
    onFieldChange: PropTypes.func,
    className: PropTypes.string,
};

FormField.defaultProps = {
    as: 'div',
    label: '',
    children: null,
    type: 'text',
    onFieldChange: () => {},
    className: '',
};

export default FormField;
