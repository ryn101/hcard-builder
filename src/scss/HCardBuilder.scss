@import 'colours';

.hcard-builder {
  flex: 1;
  flex-basis: 48%;
  padding: 40px;
  background-color: white;

  &__title {
    font-size: 24pt;
    font-weight: 800;
    margin:0 0 25px;
    color: $sophisticated-navy;
    letter-spacing: 0;

    &::first-letter {
      font-size: 25pt;
    }
  }
}
