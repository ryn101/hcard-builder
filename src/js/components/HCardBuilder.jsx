import React, { Component } from 'react';
import PropTypes from 'prop-types';
import FormSection from './FormSection';
import FormField from './FormField';
import FileUpload from './FileUpload';
import Button from './Button';
import '../../scss/HCardBuilder.scss';

class HCardBuilderComponent extends Component {
    onFieldChange = ({ target }) => {
        const { onFieldUpdate } = this.props;

        onFieldUpdate(target.id, target.value);
    };

    onAvatarFileSelected = ({ target }) => {
        const { onAvatarSelected } = this.props;

        onAvatarSelected(target.files[0]);
    };

    render() {
        const { onFormSubmission } = this.props;

        return (
            <section className="hcard-builder">
                <h1 className="hcard-builder__title">hCard Builder</h1>

                <form name="hCardForm" className="form" onSubmit={onFormSubmission}>
                    <FormSection title="Personal Details">
                        <ul className="flex-row flex-row--columns-2">
                            <FormField as="li" label="Given Name" fieldId="givenName" onFieldChange={this.onFieldChange} className="flex-row__column" />
                            <FormField as="li" label="Surname" fieldId="surname" onFieldChange={this.onFieldChange} className="flex-row__column" />
                            <FormField as="li" label="Email" fieldId="email" type="email" onFieldChange={this.onFieldChange} className="flex-row__column" />
                            <FormField as="li" label="Phone" fieldId="phone" type="tel" onFieldChange={this.onFieldChange} className="flex-row__column" />
                        </ul>
                    </FormSection>
                    <FormSection title="Address">
                        <ul className="flex-row flex-row--columns-2">
                            <FormField as="li" label="House Name or #" fieldId="houseNameOrNumber" onFieldChange={this.onFieldChange} className="flex-row__column" />
                            <FormField as="li" label="Street" fieldId="street" onFieldChange={this.onFieldChange} className="flex-row__column" />
                            <FormField as="li" label="Suburb" fieldId="suburb" onFieldChange={this.onFieldChange} className="flex-row__column" />
                            <FormField as="li" label="State" fieldId="state" onFieldChange={this.onFieldChange} className="flex-row__column" />
                            <FormField as="li" label="Postcode" fieldId="postcode" className="flex-row__column">
                                <input type="number" pattern="\d*" onChange={this.onFieldChange} />
                            </FormField>
                            <FormField as="li" label="Country" fieldId="country" onFieldChange={this.onFieldChange} className="flex-row__column" />
                        </ul>
                    </FormSection>
                    <div className="flex-row flex-row--columns-2">
                        <div className="flex-row__column form__button-wrapper">
                            <FileUpload onFileSelection={this.onAvatarFileSelected}>
                                <Button as="span">Upload Avatar</Button>
                            </FileUpload>
                        </div>
                        <div className="flex-row__column form__button-wrapper">
                            <Button type="submit" colour="blue">Create hCard</Button>
                        </div>
                    </div>
                </form>
            </section>
        );
    }
}

HCardBuilderComponent.propTypes = {
    onFieldUpdate: PropTypes.func.isRequired,
    onFormSubmission: PropTypes.func.isRequired,
    onAvatarSelected: PropTypes.func.isRequired,
};

export default HCardBuilderComponent;
