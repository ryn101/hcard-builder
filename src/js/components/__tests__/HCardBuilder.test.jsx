import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import HCardBuilder from '../HCardBuilder';
import FileUpload from '../FileUpload';
import FormField from '../FormField';

configure({ adapter: new Adapter() });

it('renders the hCard builder without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
        <HCardBuilder
            onAvatarSelected={() => {}}
            onFieldUpdate={() => {}}
            onFormSubmission={() => {}}
        />,
        div,
    );
    ReactDOM.unmountComponentAtNode(div);
});

describe('once rendered', () => {
    let hCardBuilderNode;
    const onAvatarSelected = jest.fn();
    const onFieldUpdate = jest.fn();
    const onFormSubmission = jest.fn();

    beforeEach(() => {
        hCardBuilderNode = shallow(
            <HCardBuilder
                onFormSubmission={onFormSubmission}
                onAvatarSelected={onAvatarSelected}
                onFieldUpdate={onFieldUpdate}
            />,
        );
    });

    it('renders with state correctly', () => {
        expect(hCardBuilderNode).toMatchSnapshot();
    });

    it('calls the on form submission with the correct argument', () => {
        const eventObject = {
            target: {
                files: ['file'],
            },
        };

        hCardBuilderNode.find(FileUpload).prop('onFileSelection')(eventObject);
        expect(onAvatarSelected).toBeCalledTimes(1);
        expect(onAvatarSelected).toBeCalledWith(eventObject.target.files[0]);
    });

    it('calls the on field updated with the correct argument', () => {
        const eventObject = {
            target: {
                id: 'fieldId',
                value: 'fieldValue',
            },
        };

        hCardBuilderNode.find(FormField).first().prop('onFieldChange')(eventObject);
        expect(onFieldUpdate).toBeCalledTimes(1);
        expect(onFieldUpdate).toBeCalledWith(eventObject.target.id, eventObject.target.value);
    });

    it('calls the on form submitted', () => {
        const formNode = hCardBuilderNode.find('form');
        formNode.simulate('submit', { preventDefault: () => {} });

        expect(onFormSubmission).toBeCalledTimes(1);
    });
});
