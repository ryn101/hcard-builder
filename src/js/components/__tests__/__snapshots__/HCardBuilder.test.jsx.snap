// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`once rendered renders with state correctly 1`] = `
<section
  className="hcard-builder"
>
  <h1
    className="hcard-builder__title"
  >
    hCard Builder
  </h1>
  <form
    className="form"
    name="hCardForm"
    onSubmit={[MockFunction]}
  >
    <FormSection
      title="Personal Details"
    >
      <ul
        className="flex-row flex-row--columns-2"
      >
        <FormField
          as="li"
          className="flex-row__column"
          fieldId="givenName"
          label="Given Name"
          onFieldChange={[Function]}
          type="text"
        />
        <FormField
          as="li"
          className="flex-row__column"
          fieldId="surname"
          label="Surname"
          onFieldChange={[Function]}
          type="text"
        />
        <FormField
          as="li"
          className="flex-row__column"
          fieldId="email"
          label="Email"
          onFieldChange={[Function]}
          type="email"
        />
        <FormField
          as="li"
          className="flex-row__column"
          fieldId="phone"
          label="Phone"
          onFieldChange={[Function]}
          type="tel"
        />
      </ul>
    </FormSection>
    <FormSection
      title="Address"
    >
      <ul
        className="flex-row flex-row--columns-2"
      >
        <FormField
          as="li"
          className="flex-row__column"
          fieldId="houseNameOrNumber"
          label="House Name or #"
          onFieldChange={[Function]}
          type="text"
        />
        <FormField
          as="li"
          className="flex-row__column"
          fieldId="street"
          label="Street"
          onFieldChange={[Function]}
          type="text"
        />
        <FormField
          as="li"
          className="flex-row__column"
          fieldId="suburb"
          label="Suburb"
          onFieldChange={[Function]}
          type="text"
        />
        <FormField
          as="li"
          className="flex-row__column"
          fieldId="state"
          label="State"
          onFieldChange={[Function]}
          type="text"
        />
        <FormField
          as="li"
          className="flex-row__column"
          fieldId="postcode"
          label="Postcode"
          onFieldChange={[Function]}
          type="text"
        >
          <input
            onChange={[Function]}
            pattern="\\\\d*"
            type="number"
          />
        </FormField>
        <FormField
          as="li"
          className="flex-row__column"
          fieldId="country"
          label="Country"
          onFieldChange={[Function]}
          type="text"
        />
      </ul>
    </FormSection>
    <div
      className="flex-row flex-row--columns-2"
    >
      <div
        className="flex-row__column form__button-wrapper"
      >
        <FileUpload
          id="avatarFileUpload"
          labelClass="file-upload__label"
          onFileSelection={[Function]}
        >
          <Button
            as="span"
            colour="grey"
          >
            Upload Avatar
          </Button>
        </FileUpload>
      </div>
      <div
        className="flex-row__column form__button-wrapper"
      >
        <Button
          as="button"
          colour="blue"
          type="submit"
        >
          Create hCard
        </Button>
      </div>
    </div>
  </form>
</section>
`;
